Add factual rewards to synthetic data generation and enable online bandit simulations

Braamling opened the pull request:
### Why
As an applied researcher on applications of bandits, I want to be able to compare various bandits algorithms in simulation. 

### About the change
The current implementation of the bandit simulations use a synthetic bandits feedback log under the assumption of some logging policy. While training an online policy (eg. epsilon greedy), the policy is only updated when the taken action of the online policy is identical to the action taken by the (synthetic) logging policy. This might make sense when we want to optimise the bandit policy using logged data, however, this does not seem sensible when we are using a simulated data anyway. 

For this reason we've made the following changes to simulation aspect of the framework
* A `factual_reward` field was added to the `SyntheticBanditDataset`, which is a full-information sampled reward from the `expected_reward` in the same dataset. 
* `run_bandit_simulation` is modified to use `factual_reward` to update bandit policies. Updates happen each trial, regardless of whether the action is identical to the logged action.

### Remarks
I understand that with this change I might break some explicit assumptions that were made for the `run_bandit_simulation` method. I'm happy to be advised on how my change breaks these assumptions and how best to proceed.

Braamling commented on obp/dataset/synthetic_slate.py:
```diff
@@ -965,7 +965,7 @@ def calc_ground_truth_policy_value(
         else:
             n_batch = (
                 n_rounds * n_enumerated_slate_actions * self.len_list - 1
-            ) // 10**7 + 1
```
I think Black has updated its syntax...

Braamling commented:
@usaito I'd love to get your feedback on this MR :)

usaito commented:
@Braamling Thank you for your contribution! and let me get back to you sometime next week.

usaito commented:
@Braamling  Thank you so much again for your contributions! I've checked your implementations and actually found them to be controversial in the following aspects. In my understanding, it seems that you are trying to expand the package by enabling it to perform some simulation of bandit algorithms in the online environment. However, our package focuses more on offline evaluation of batch bandit policies and online bandit policies, which is actually greatly different from implementing online simulation (like OpenAi gym). This is why: 

- `obp.dataset.SyntheticBanditDataseat` generates the logged data in the batch manner, and does not include "factual_rewards" (according to your terminology, but actually, this is not "factual," in my opinion, just an independent minor point).

- `obp.simulator.run_bandit_simulation` uses only partial observations in the logged data depending on how the actions are chosen by new and old bandit policies, so to imitate the situation where we use only the logged data to evaluate online policies (https://arxiv.org/abs/1003.5956)

Although it makes sense to additionally implement some features enabling the online simulation of online bandits (rather than their offline evaluation), I would suggest doing this by

- implementing some new classes providing online simulation environment (like a gym environment)  rather than modifying the existing `obp.dataset.SyntheticBanditDataseat`, which focuses on generating **offline** logged bandit data

- implementing a new function to run online simulations of online bandits rather than modifying the existing `obp.simulator.run_bandit_simulation`, which focuses on evaluating online bandit policies **using only offline logged data** (see https://github.com/st-tech/zr-obp/blob/master/examples/quickstart/online.ipynb as a reference)

But again, let me just clarify that OBP is not mainly motivated to implement or simulate online bandit policies in the online environment (which may be implemented with many other packages focusing on standard online bandits).

Anyway, thank you again for your contributions, and I'm happy to hear your comments and take any further relevant questions!

Braamling commented:
@usaito, thank you for your clear response!

I've been truly impressed with the ease of use and quality of code with OBP so far. For us, it has been a great facilitator to experiment with both simulations of online bandits and off-policy techniques.  We're working on a workshop paper that describes two more extensions for OBP based on the proposed online simulations which helps experimenting with concepts such as reward delay, concept drift and comparing policies trained on-policy and through logged data. 

I agree that keeping the concepts of simulations for replay methods and online simulations separate makes more sense. It should not be too complex to separate out the new code, while keeping the code duplication to a minimum.

My proposal based on your suggestions would be to create a new class named `BanditSimulator` that uses the same API as the offline `SyntheticBanditDataset`, but instead returns the rewards of all arms (Instead of the "factual" rewards). This means that the `behavior_policy_function`, `beta`, `n_deficient_actions` will be omitted form this class. The existing `SyntheticBanditDataset` will remain as it is currently written in the main branch of the OBP repo. 

For the simulation, I would propose to rename the current `run_bandit_simulation` to `run_bandit_replay` and still make my proposed changes to `run_bandit_simulation`. This new method can either retrieve a `BanditSimulator` object or a batch of data from the `BanditSimulator` to run a simulation.  I might be mixing up terminology here, so please feel free to provide another suggestion, but to me it feels that the current use of the term simulation is different from what I would expect the underlying method to do (replay on a logged dataset). If you agree to the proposed change, I will of course make sure that the rest of the existing code and notebooks are refactored to use the new `run_bandit_replay` method.

The advantage of these changes could be that use can use the BanditSimulator to create more extensive behavioural experiments that can then be evaluated with the existing OPE and OPO tools available in OBP. 

I want to thank you again for the amazing work on the OBP framework and hope that we can find more ways to contribute while we continue our work on Bandit algorithms and OPE at Booking.com

usaito commented:
@Braamling Great to hear your plan for the workshop paper, and thank you for your suggestions! I agree that expanding the package by implementing an online simulation environment makes a lot of sense, although the main motivation of the package will remain on the offline side. The followings are some of my thoughts on your suggestions.


> My proposal based on your suggestions would be to create a new class named BanditEnvironmentSimulator that uses the same API as the offline SyntheticBanditDataset, but instead returns the rewards of all arms (Instead of the "factual" rewards).

I agree to implement `BanditEnvironmentSimulator` (or how about `SimulationBanditEnvironment`? which I feel is a bit better) as a new class. However, I'd suggest implementing this new class in a different way compared to `SyntheticBanditDataset`, because online and offline bandits assume very different data generating processes (DGP), and I want each class to precisely follow the corresponding DGPs. In offline bandit evaluation or learning, we have some batch logged datasets collected by a logging policy, so `SyntheticBanditDataset` generates batch logged data in advance to OPL and OPL. However, in online bandits, we observe the data one by one in response to a bandit policy's output/choice. So, for example, I'd expect..

```
cumulative_reward = 0 # performance measure
env = SimulationBanditEnvironment(...
for t in range(1000):
     context = env.sample_context()
     action = policy.select_action(context)
     reward, expected_reward = env.sample_reward(context, action, need_expected_reward=True)
     policy.update_params(action, reward, context)
     cumulative_reward += expected_reward
```
and it would be great to have your thoughts on this. And, in my understanding, the above process will be the inner-working of the new `run_bandit_simulation` function in the following.


> This means that the behavior_policy_function, beta, n_deficient_actions will be omitted from this class. The existing SyntheticBanditDataset will remain as it is currently written in the main branch of the OBP repo.

Yes, I totally agree with this point. Thank you for the clarification! 


> For the simulation, I would propose to rename the current run_bandit_simulation to run_bandit_replay and still make my proposed changes to run_bandit_simulation.

Thank you for the suggestion! My understanding is.. 

- first you'll simply rename `run_bandit_simulation` to `run_bandit_replay` without any changes in the original implementation, I mean, `run_bandit_replay` will be used only as an offline evaluation method
- and then you'll implement a new `run_bandit_simulation` function, which is to facilitate simulations of online bandits in an online environment 

It would be great if you could see if my understanding is correct. If it's correct, I agree with the procedure, and `run_bandit_replay` seems to be a great correction. I'd expect that the new `run_bandit_simulation` function will take a `SimulationBanditEnvironment` object to run a simulation given my suggestions above. 
 
Thank you again for your suggestions, and very happy to continue the discussion in this direction!

Braamling commented:
Thanks @usaito,

I've taken your suggestions and started on a refactor. It has become a bit of a big MR, so let me know if you want me to split it up (see #177).

The only difference of my implementation and your latest comment are:
- Instead of a `run_bandit_simulation` method, I've implemented a class named `BanditPolicySimulator`. This class keeps track of what happens in the simulation, which can afterward be used as a logged dataset for off-policy. I think this nicely connects the online and offline world in OBP.
- Instead of doing an interaction with the environment as you suggested, I sample a `BanditRound` which contains sampled values of all arms (full information). The nice thing about this is that the same dataset can be replayed for multiple algorithms and that we can be much more efficient computationally. 
- I've also added the drift and delay functionality in the same MR, this was a lot easier for the refactoring. Let me know if you have concerns about keeping this in one MR.

Let me know what you think on a high level of these changes (I still need to do some final cleaning up, which I will do this week)
